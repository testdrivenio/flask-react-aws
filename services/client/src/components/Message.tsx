import React from "react";
import {
  Alert,
  AlertIcon,
  AlertTitle,
  CloseButton,
  Box,
  Container,
} from "@chakra-ui/react";

interface MessageProps {
  messageType: "info" | "warning" | "success" | "error";
  messageText: string;
  onClose?: () => void;
}

const Message: React.FC<MessageProps> = ({
  messageType,
  messageText,
  onClose,
}) => {
  return (
    <Container maxW="container.2xl" px={4}>
      <Alert
        status={messageType}
        variant="solid"
        mt={4}
        borderRadius="md"
        display="flex"
        alignItems="center"
      >
        <AlertIcon />
        <Box flex="1">
          <AlertTitle>{messageText}</AlertTitle>
        </Box>
        {onClose && (
          <CloseButton
            onClick={onClose}
            ml={2}
            size="sm"
            color="white"
            _hover={{ bg: "whiteAlpha.300" }}
          />
        )}
      </Alert>
    </Container>
  );
};

export default Message;
