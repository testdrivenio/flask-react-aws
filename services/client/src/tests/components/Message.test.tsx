import React from "react";
import { render, screen, cleanup } from "../test-utils";
import { describe, it, expect, afterEach, vi } from "vitest";
import { ChakraProvider } from "@chakra-ui/react";
import Message from "../../components/Message";

const renderWithChakra = (component: React.ReactElement) => {
  return render(<ChakraProvider>{component}</ChakraProvider>);
};

describe("Message Component", () => {
  afterEach(() => {
    cleanup();
    vi.restoreAllMocks();
  });

  it("renders a success message", () => {
    const props = {
      messageType: "success" as const,
      messageText: "Operation successful!",
    };
    renderWithChakra(<Message {...props} />);

    const alert = screen.getByRole("alert");
    expect(alert).toBeDefined();
    expect(alert.getAttribute("data-status")).toBe("success");
    expect(screen.getByText("Operation successful!")).toBeDefined();
  });

  it("renders an error message", () => {
    const props = {
      messageType: "error" as const,
      messageText: "An error occurred.",
    };
    renderWithChakra(<Message {...props} />);

    const alert = screen.getByRole("alert");
    expect(alert).toBeDefined();
    expect(alert.getAttribute("data-status")).toBe("error");
    expect(screen.getByText("An error occurred.")).toBeDefined();
  });

  it("renders an info message", () => {
    const props = {
      messageType: "info" as const,
      messageText: "Here is some information.",
    };
    renderWithChakra(<Message {...props} />);

    const alert = screen.getByRole("alert");
    expect(alert).toBeDefined();
    expect(alert.getAttribute("data-status")).toBe("info");
    expect(screen.getByText("Here is some information.")).toBeDefined();
  });

  it("renders a close button when onClose prop is provided", () => {
    const onCloseMock = vi.fn();
    const props = {
      messageType: "info" as const,
      messageText: "Closable message",
      onClose: onCloseMock,
    };
    renderWithChakra(<Message {...props} />);

    const closeButton = screen.getByRole("button");
    expect(closeButton).toBeDefined();
  });

  it("does not render a close button when onClose prop is not provided", () => {
    const props = {
      messageType: "info" as const,
      messageText: "Non-closable message",
    };
    renderWithChakra(<Message {...props} />);

    const closeButton = screen.queryByRole("button");
    expect(closeButton).toBeNull();
  });
});

describe("Message Component Snapshots", () => {
  it("matches snapshot for success message", () => {
    const props = {
      messageType: "success" as const,
      messageText: "Operation successful!",
    };
    const { asFragment } = renderWithChakra(<Message {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for error message", () => {
    const props = {
      messageType: "error" as const,
      messageText: "An error occurred.",
    };
    const { asFragment } = renderWithChakra(<Message {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for info message", () => {
    const props = {
      messageType: "info" as const,
      messageText: "Here is some information.",
    };
    const { asFragment } = renderWithChakra(<Message {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for message with close button", () => {
    const props = {
      messageType: "info" as const,
      messageText: "Closable message",
      onClose: () => {},
    };
    const { asFragment } = renderWithChakra(<Message {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
