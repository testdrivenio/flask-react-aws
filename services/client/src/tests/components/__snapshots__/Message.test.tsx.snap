// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Message Component Snapshots > matches snapshot for error message 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-bl6ftg"
  >
    <div
      class="chakra-alert css-1omal6u"
      data-status="error"
      role="alert"
    >
      <span
        class="chakra-alert__icon css-14ogjxt"
        data-status="error"
      >
        <svg
          class="chakra-icon css-qk6lof"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div
        class="css-1rr4qq7"
      >
        <div
          class="chakra-alert__title css-tidvy5"
          data-status="error"
        >
          An error occurred.
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`Message Component Snapshots > matches snapshot for info message 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-bl6ftg"
  >
    <div
      class="chakra-alert css-1i1aypw"
      data-status="info"
      role="alert"
    >
      <span
        class="chakra-alert__icon css-14ogjxt"
        data-status="info"
      >
        <svg
          class="chakra-icon css-qk6lof"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div
        class="css-1rr4qq7"
      >
        <div
          class="chakra-alert__title css-tidvy5"
          data-status="info"
        >
          Here is some information.
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`Message Component Snapshots > matches snapshot for message with close button 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-bl6ftg"
  >
    <div
      class="chakra-alert css-1i1aypw"
      data-status="info"
      role="alert"
    >
      <span
        class="chakra-alert__icon css-14ogjxt"
        data-status="info"
      >
        <svg
          class="chakra-icon css-qk6lof"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div
        class="css-1rr4qq7"
      >
        <div
          class="chakra-alert__title css-tidvy5"
          data-status="info"
        >
          Closable message
        </div>
      </div>
      <button
        aria-label="Close"
        class="css-cm0zi7"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`Message Component Snapshots > matches snapshot for success message 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-bl6ftg"
  >
    <div
      class="chakra-alert css-1ky8msv"
      data-status="success"
      role="alert"
    >
      <span
        class="chakra-alert__icon css-14ogjxt"
        data-status="success"
      >
        <svg
          class="chakra-icon css-qk6lof"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,0A12,12,0,1,0,24,12,12.014,12.014,0,0,0,12,0Zm6.927,8.2-6.845,9.289a1.011,1.011,0,0,1-1.43.188L5.764,13.769a1,1,0,1,1,1.25-1.562l4.076,3.261,6.227-8.451A1,1,0,1,1,18.927,8.2Z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div
        class="css-1rr4qq7"
      >
        <div
          class="chakra-alert__title css-tidvy5"
          data-status="success"
        >
          Operation successful!
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
